[uArch][XeGPU] Add XeGPU uArch definition.

mshahneo opened the pull request:
The uArch infrastructure provides:
- A set data structures to represent, uArch and it's necessary components (e.g., instructions, register-files, caches).
- A set of utility interfaces that are common to a family of ops (e.g., mma ops, 2DBlockIO ops). The implementation of these interfaces are provided by the specific instructions. Each family of ops provides these 5 common APIs. However, some family of ops may have more utility APIs. The common 5 APIs are:
	- getSupportedShapes
	- getSupportedTypes
	- checkSupportedShapesAndTypes
	- checkSupportedTypes
	- validate

Add support for PVC and BMG architectures.
Add support for DPAS instruction.

mshahneo committed:
[uArch][XeGPU] Add XeGPU uArch definition.

The uArch infrastructure provides:
- A set data structures to represent, uArch and it's necessary components
  (e.g., instructions, register-files, caches).
- A set of utility interfaces that are common to a family of ops
  (e.g., mma ops, 2DBlockIO ops). The implementation of these interfaces
  are provided by the specific instructions. Each family of ops provides
  these 5 common APIs. However, some family of ops may have more
  utility APIs. The common 5 APIs are:
	- getSupportedShapes
	- getSupportedTypes
	- checkSupportedShapesAndTypes
	- checkSupportedTypes
	- validate

Add support for PVC and BMG architectures.
Add support for DPAS instruction.

adam-smnk reviewed (commented):
High-level question: is there any info that gets pulled at runtime?
I imagine most of these hardware properties don't really change. Can't all this be static compile time info?

mshahneo committed:
Address review comments.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,266 @@
+///
+//
+//===----------------------------------------------------------------------===//
+#ifndef MLIR_DIALECT_XEGPU_UARCH_BASE_H
```
Done.

mshahneo commented on mlir/lib/Dialect/XeGPU/uArch/IntelGpuXe2.cpp:
```diff
@@ -0,0 +1,197 @@
+#include "mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/Support/YAMLTraits.h"
```
Removed.

mshahneo commented on mlir/lib/Dialect/XeGPU/uArch/IntelGpuXe2.cpp:
```diff
@@ -0,0 +1,197 @@
+#include "mlir/IR/BuiltinTypes.h"
+#include "llvm/Support/YAMLTraits.h"
+#include <algorithm>
+#include <iostream>
```
Removed.

mshahneo commented on mlir/lib/Dialect/XeGPU/IR/XeGPUDialect.cpp:
```diff
@@ -35,6 +36,14 @@ void XeGPUDialect::initialize() {
 #include <mlir/Dialect/XeGPU/IR/XeGPUAttrs.cpp.inc>
       >();
+
+  // Populate the uArchMap with the supported target devices
```
I can remove this from the dialect initializer.
I kept it here because this way, the uArch map is populated once the dialect is loaded. Any pass or the dialect can use it from there on. But other option can also work, where we a pass can initialize the uArch map and use it afterwards.

Any specific reason why it can't be in dialect initializer?

adam-smnk replied:
Right now, uArch info is not strictly necessary for the dialect i.e., you can still do a lot like transforming ops and running generic transforms without pulling in the hardware info. It makes me "pay the price" for sth I don't need.
I can see that moving forward this might change as a lot of passes might want to use it. But that's a bit premature decision.

Also, AFAIK the whole runtime initialization is the reason why you had to introduce mutexes into table builders. It adds extra complexity both design and runtime without a clear payoff.

mshahneo commented:
> High-level question: is there any info that gets pulled at runtime? I imagine most of these hardware properties don't really change. Can't all this be static compile time info?

Nothing in particular, all are static info. I actually tried to follow the MLIR way. Since it is like an utility, I tried to follow the utility functions in different dialects. But I am open to change.

akroviakov commented on mlir/lib/Dialect/XeGPU/uArch/IntelGpuXe2.cpp:
```diff
@@ -0,0 +1,172 @@
+}
+
+std::vector<mlir::Type>
+DPASInstruction::getSupportedTypes(MLIRContext &context,
```
What is the motivating use case behind this method?

akroviakov reviewed (commented):
The choice of runtime vs static allocations and data structures can be optimized in the subsequent iterations.
The "string vs enum" is more of a structure/API decision, so it can be discussed in this PR. Is there any particular reason to favor strings over enums?
 
The uarch has some non-obvious structure (`no_of_component` for _lower_ level, hierarchy level inclusion in cache info struct). Without any motivating example these seem overcomplicated and are likely to be replaced/moved anyway when a concrete use case appears. 

How about a lightweight definition first (e.g., excluding `num_matrix_units`), but with at least one practical usage in any part of xegpu (e.g., dpas)? As you noted, there are quite a number of fields that can be added, so why not let use cases decide which fields should actually be added.

adam-smnk reviewed (commented):
I still have concerns when it comes to wrapping fully static data into runtime structures.
It adds both complexity to the design and runtime overhead.

> The choice of runtime vs static allocations and data structures can be optimized in the subsequent iterations.

Pragmatically speaking, this rarely happens.

Design choices aside, overall info and utilities provided look useful.
The exact APIs can be refined once we start integrating uArch info into passes. Then we'll see what works and what's missing.

It might be better to just take it for a test drive once a general cleanup is done.

mshahneo committed:
Address review comments.

Simplify the design:
- Remove uArchHierarchyComponent

LLVMize names.
Replace String usage with enum whenever possible.

mshahneo committed:
Address review comments.

Simplify design:
  - Remove dialect initialization and necessary mechanism.

mshahneo committed:
Address review comments.

Move all the implementation to the .h file.
Move uArchInterfaces to uArchBase.

mshahneo committed:
Address review comments.

Use LLVM data structures whenever possible.

mshahneo committed:
Add/Remove some spacings.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,183 @@
+                {/* restrictions */}        // Optional: empty
+        ) {
+    // Initialize uArchHierarchy
+    this->uArch_hierarchy.push_back(uArchHierarchyComponent("thread", 0));
```
Removed it to make the design simpler.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,266 @@
+
+struct Instruction {
+  // @TODO: Add more fields as needed
+  Instruction(std::string name, std::string desc)
```
Ah, missed it, fixed.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,183 @@
+  }
+};
+
+struct Xe2Plus : public uArch {
```
Removed the extra namespaces.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,183 @@
+                {/* instructions */}        // Optional: empty
+        ) {
+    // Initialize uArchHierarchy
+    this->uArch_hierarchy.push_back(uArchHierarchyComponent("thread", 0));
```
Removed them for now. Will think about using interfaces when we bring them back.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,183 @@
+#include <vector>
+
+namespace mlir {
+namespace xegpu {
```
I understand the issue.
The problem with XeVM dialect is that it is part of the target dialects and lives with other leaf dialects. Didn't want to convolute that.

mshahneo committed:
Address review comments.

Jianhui-Li commented on mlir/include/mlir/Dialect/XeGPU/uArch/IntelGpuXe2.h:
```diff
@@ -0,0 +1,309 @@
+namespace mlir {
+namespace xegpu {
+namespace uArch {
+struct XeCoreInfo {
```
nit: suggest move to uArchBase.h

Jianhui-Li reviewed (approved):
LGTM.

akroviakov reviewed (approved):
LGTM

mshahneo committed:
Address review comments.

mshahneo committed:
Address review comments.

mshahneo commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -0,0 +1,266 @@
+  std::string getDescription() { return description; }
+  InstructionScopeEnum getScope() { return scope; }
+
+protected:
```
Done.

mshahneo committed:
Fix a small compile error.

rolfmorel commented on mlir/include/mlir/Dialect/XeGPU/uArch/uArchBase.h:
```diff
@@ -23,179 +23,167 @@
+  // @TODO: Add more instructions as needed
+};
 
+llvm::StringRef toString(InstructionKind name) {
```
Nit: would this make more sense as a method?

(Similarly, parseInstructionKind _could_ be a static method.)

Matter of taste though. If there's precedence in the codebase, please ignore this comment.

mshahneo committed:
Address review comments.

mshahneo committed:
Address review comments.

Make the some helper funcs to static methods.

adam-smnk reviewed (approved):
Sorry for the delay. Slowly catching up with PRs 😊

Overall structure looks really good. Thanks for the extensive rework 👍
I left a few minor cleanup remarks. Otherwise, good to go.
